yew-macro/src/html_tree/html_tag/mod.rs

```diff
@@ -203,23 +307,22 @@ struct HtmlTagOpen {
     gt: Token![>],
 }
 
-impl PeekValue<TagName> for HtmlTagOpen {
-    fn peek(cursor: Cursor) -> Option<TagName> {
+impl PeekValue<TagKey> for HtmlTagOpen {
+    fn peek(cursor: Cursor) -> Option<TagKey> {
         let (punct, cursor) = cursor.punct()?;
         (punct.as_char() == '<').as_option()?;
 
-        let (name, cursor) = TagName::peek(cursor)?;
-        if name.to_string() == "key" {
-            let (punct, _) = cursor.punct()?;
-            if punct.as_char() == '=' {
-                None
+        let (tag_key, cursor) = TagName::peek(cursor)?;
+        if let TagKey::Lit(name) = &tag_key {
+            if name.to_string() == "key" {
+                let (punct, _) = cursor.punct()?;
+                (punct.as_char() != '=').as_option()?;
```

Comment:
I'm kind of confused by this check here...
I couldn't figure out why key was given a special case in the peek method of all places so I tried removing it and ran the tests. Here's what the errors look like when "key" doesn't get special treatment:

**1.**
```diff
- error: unexpected end of input, expected expression
+ error: unexpected token
     |
- 11 |     html! { <key=></>}
-    |     ^^^^^^^^^^^^^^^^^^
+ 11 |     html! { <key=></>}
+    |                 ^^
```
**2.**
```diff
- error: this opening tag has no corresponding closing tag
+ error: unexpected token
     |
- 12 |     html! { <key="key".to_string()>invalid</key>}
-    |             ^^^^^^^^^^^^^^^^^^^^^^^
+ 12 |     html! { <key="key".to_string()>invalid</key>}
+    |                 ^
```

---

In both cases it seems to me that the current error message isn't particularly helpful. Of course removing the check doesn't really make it any better though the first error now points to the appropriate location.
Should this check be done in the `parse` method so it can return a proper error message for all tag names that look like attribute assignments?
It's also very possible that I'm just missing the point here. Is there some place where it's important that tags in the form of `<key={...}>` not be detected as opening tags?

Reply:
Interesting, I agree that the error messages could be clearer here.

This is checking for keyed fragments. So just like `<div key="123">` can be keyed, so can fragments like so:
```rust
html! {
  <>
    <key="a">..</>
    <key="b">..</>
  </>
}
```

Reply:
And because fragments need to be treated as `HtmlList` we make sure that `HtmlTag` doesn't accidentally detect them? That makes sense to me - Thanks for clarifying.
I also failed to notice that the `html-list-pass` test no longer passes when the check is removed.

The error message for the second case is fine as it is IMO. Perhaps it could be changed to read "this opening fragment [...]" instead of "tag". When glossing over the code `<key="key".to_string()>invalid</key>` it's tempting to think that `</key>` should actually close the opening tag and writing "fragment" might help here.

<br/>

I tracked down the origin of the first case and this error message is emitted by `HtmlProp` when there's no expression after the equals sign. What I don't understand is why it seems to lose the proper location of the error...
I modified the `HtmlProp` method so it checks for an empty input stream after the equals sign.
The error now reads:
```rust
error: expected an expression following this equals sign
   |
11 |     html! { <key=></>}
   |                 ^^
```
I think this doesn't cover all cases but it's certainly an improvement.
The only other way I could think of would be to `map_err` the result of the expression parsing itself but then we lose all the details for errors within the expression itself.

Reply:
> And because fragments need to be treated as HtmlList we make sure that HtmlTag doesn't accidentally detect them? 

Yes, exactly!

> Perhaps it could be changed to read "this opening fragment [...]" instead of "tag".

Yeah, agreed. That would be clearer

> I think this doesn't cover all cases but it's certainly an improvement.

Agreed, this is fine for now 👍